tsdb/head_append.go

```diff
@@ -1322,17 +1331,84 @@ func (s *memSeries) appendPreprocessor(t int64, e chunkenc.Encoding, o chunkOpts
 	return c, true, chunkCreated
 }
 
+// histogramsAppendPreprocessor takes care of cutting new histogram chunks and m-mapping old ones. Histogram chunks are
+// cut based on their size in bytes.
+// It is unsafe to call this concurrently with s.iterator(...) without holding the series lock.
+// This should be called only when appending data.
+func (s *memSeries) histogramsAppendPreprocessor(t int64, e chunkenc.Encoding, o chunkOpts) (c *memChunk, sampleInOrder, chunkCreated bool) {
+	c = s.headChunks
+
+	if c == nil {
+		if len(s.mmappedChunks) > 0 && s.mmappedChunks[len(s.mmappedChunks)-1].maxTime >= t {
+			// Out of order sample. Sample timestamp is already in the mmapped chunks, so ignore it.
+			return c, false, false
+		}
+		// There is no head chunk in this series yet, create the first chunk for the sample.
+		c = s.cutNewHeadChunk(t, e, o.chunkRange)
+		chunkCreated = true
+	}
+
+	// Out of order sample.
+	if c.maxTime >= t {
+		return c, false, chunkCreated
+	}
+
+	if c.chunk.Encoding() != e {
+		// The chunk encoding expected by this append is different than the head chunk's
+		// encoding. So we cut a new chunk with the expected encoding.
+		c = s.cutNewHeadChunk(t, e, o.chunkRange)
+		chunkCreated = true
+	}
+
+	numSamples := c.chunk.NumSamples()
+	targetBytes := chunkenc.TargetBytesPerHistogramChunk
+	numBytes := len(c.chunk.Bytes())
+
+	if numSamples == 0 {
+		// It could be the new chunk created after reading the chunk snapshot,
+		// hence we fix the minTime of the chunk here.
+		c.minTime = t
+		s.nextAt = rangeForTimestamp(c.minTime, o.chunkRange)
+	}
+
```

Comment:
I think we can implement the corresponding (conservative) behavior as we know it from XOR chunks  in the following:

If `s.histogramChunkHasComputedEndTime` is false here, we can simply set `nextChunkRangeStart` to the value of `s.nextAt` (because it was either set at chunk creation, or it was set when loading from the snapshot - either way, we want to end the chunk at that time, no matter what). And we must not set `nextChunkRangeStart` below, see note there.

Reply:
Nice, yeah I think that works! I'll make the change.